crates/symbolicator/src/services/cacher.rs

```diff
@@ -372,17 +406,148 @@ impl<T: CacheItemRequest> Cacher<T> {
+    use std::time::{Duration, Instant};
+
+    use futures::compat::Future01CompatExt;
+    use futures::Future;
+
+    use crate::config::{CacheConfig, CacheConfigs};
+    use crate::test::{self, tempdir};
+
+    use super::*;
+
+    /// A tokio 0.1 compat layer to sleep the task for the given time.
+    ///
+    /// This is necessary because the [`Cacher`] still uses a tokio 0.1 spawn internally.
+    fn sleep(millis: u64) -> impl Future<Output = ()> {
+        let when = Instant::now() + Duration::from_millis(millis);
+        tokio01::timer::Delay::new(when)
+            .compat()
+            .unwrap_or_else(|_| ())
+    }
+
+    #[derive(Clone, Default)]
+    struct TestCacheItem {
+        computations: Arc<AtomicUsize>,
+    }
+
+    impl CacheItemRequest for TestCacheItem {
+        type Item = String;
+
+        type Error = std::io::Error;
+
+        const VERSIONS: CacheVersions = CacheVersions {
+            current: 1,
+            fallbacks: &[0],
+        };
+
+        fn get_cache_key(&self) -> CacheKey {
+            CacheKey {
+                cache_key: String::from("some_cache_key"),
+                scope: Scope::Global,
+            }
+        }
+
+        fn compute(&self, path: &Path) -> BoxedFuture<Result<CacheStatus, Self::Error>> {
+            self.computations.fetch_add(1, Ordering::SeqCst);
+
+            let path = path.to_owned();
+            Box::pin(async move {
+                sleep(50).await;
+
+                std::fs::write(path, "some new cached contents")?;
+                Ok(CacheStatus::Positive)
+            })
+        }
+
+        fn load(
+            &self,
+            _scope: Scope,
+            _status: CacheStatus,
+            data: ByteView<'static>,
+            _path: CachePath,
+        ) -> Self::Item {
+            std::str::from_utf8(data.as_slice()).unwrap().to_owned()
+        }
+    }
+
+    #[tokio::test]
+    async fn test_cache_fallback() {
+        test::setup();
+
+        let cache_dir = tempdir().path().join("test");
+        std::fs::create_dir_all(cache_dir.join("global")).unwrap();
+        std::fs::write(
+            cache_dir.join("global/some_cache_key"),
+            "some old cached contents",
+        )
+        .unwrap();
+
+        let cache = Cache::from_config(
+            "test",
+            Some(cache_dir),
+            None,
+            CacheConfig::from(CacheConfigs::default().derived),
+        )
+        .unwrap();
+        let cacher = Cacher::new(cache);
+
+        let request = TestCacheItem::default();
+
+        test::spawn_compat(move || async move {
+            let first_result = cacher.compute_memoized(request.clone()).await;
+            assert_eq!(first_result.unwrap().as_str(), "some old cached contents");
+
+            let second_result = cacher.compute_memoized(request.clone()).await;
+            assert_eq!(second_result.unwrap().as_str(), "some old cached contents");
+
+            sleep(70).await;
```

Comment:
I'm worried this will be flaky.

Could you use an atomic bool global to signal that the computation is ready instead?  That way you can set that atomic bool to true here instead of sleeping.

Reply:
That would require a condvar or a channel probably, which is a bit annoying to set up.

I admit this might be flaky as it depends on timing, but I explicitly want to assert side effects here in this test. I will adjust the timers a bit to reduce the potential flakyness.

Reply:
Yes, convar should work, or atomic bool in a spinlock loop (sleep short bits).  it's only a test.  I still think this would be more reliable than the random sleeps.